internal/controllers/gateway/httproute_controller.go

```diff
@@ -699,3 +751,74 @@ func updateAcceptedConditionInRouteParentStatus(
 	}
 	return changed
 }
+
+// TODO: extract method for different types of *Routes to update conditions.
+func (r *HTTPRouteReconciler) ensureParentsProgrammedCondition(
+	ctx context.Context,
+	httproute *gatewayv1beta1.HTTPRoute,
+	gateways []supportedGatewayWithCondition,
+	conditionStatus metav1.ConditionStatus,
+	conditionReason gatewayv1beta1.RouteConditionReason,
+	conditionMessage string,
+) (bool, error) {
+	// map the existing parentStatues to avoid duplications
+	parentStatuses := make(map[string]*gatewayv1beta1.RouteParentStatus)
+	for _, existingParent := range httproute.Status.Parents {
+		namespace := httproute.Namespace
+		if existingParent.ParentRef.Namespace != nil {
+			namespace = string(*existingParent.ParentRef.Namespace)
+		}
+		existingParentCopy := existingParent
+		var sectionName string
+		if existingParent.ParentRef.SectionName != nil {
+			sectionName = string(*existingParent.ParentRef.SectionName)
+		}
+		parentStatuses[fmt.Sprintf("%s/%s/%s", namespace, existingParent.ParentRef.Name, sectionName)] = &existingParentCopy
+	}
```

Comment:
All these blocks could be replaced by what's introduced in #3389. (Actually a coincidence I worked on this 🙃 )

Reply:
updated, and the key of map has been changed.